Implement "exports" proposal

jkrems opened the pull request:
This implements the package.json "exports" proposal (https://github.com/jkrems/proposal-pkg-exports/) with the following features:

* When `"exports"` are defined for a package (`{ "exports": { "./x": "./x.js" }`), resolving that package `import 'pkg/x'`, will resolve the exports path (`/path/to/pkg/x.js`).
* When a subpath does not match `"exports"`, an error is thrown (see test case).
* Support for `"exports": false` to indicate a package has no exports at all.
* Support for folder exports (`{ "./folder/": "./other-folder/" }`).

This reflects the consensus from this week's modules working group to move forward with `exports` support in core.

Items to be addressed before this feature is considered "fully done":

* Support `exports` in the `require` loader.
* Consider key / target validations (if any) and provide error messages where appropriate.
* Make call on support for differential exports (array syntax in import maps).
* Make call on support for package-internal entries (e.g. an `imports` field in `package.json`).

The working group's recommendation was to address these items as follow-ups to keep the changes (and discussions) focussed on one thing at a time.

See: https://github.com/jkrems/proposal-pkg-exports/issues/36

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

jkrems commented:
/cc @nodejs/modules-active-members

GeoffreyBooth commented on doc/api/esm.md:
````diff
@@ -216,6 +216,61 @@ a package would be accessible like `require('pkg')` and `import
+```
+
+If a package has no exports, setting `"exports": false` can be used instead of
+`"exports": {}` to indicate the package does not intent for submodules to be
````
```suggestion
`"exports": {}` to indicate the package does not intend for submodules to be
```

ljharb commented:
This doesn't appear to be behind a flag?

jkrems commented:
Yeah, we punted on the flag, argument being that the ESM part at least already is guarded by the `--experimental-modules` flag. If there's a concern about landing the ESM part without a flag, we can definitely add a flag (e.g. `experimental-exports`).

ljharb commented:
I’m concerned about later adding the cjs part without a preexisting flag for the exports functionality.

GeoffreyBooth commented:
Can we just put all of `"exports"`, both the ESM and CommonJS sides of it, under the `--experimental-modules` flag? I think `"exports"` should be released unflagged along with the overall new modules implementation being unflagged, and keeping it all under the same flag would ensure that.

ljharb commented:
No - i think the cjs support for exports absolutely needs to be releasable prior to/separate from the ESM implementation. In the last meeting i very explicitly said i did not think this feature should be coupled to releasing ESM.

addaleax commented on src/env.h:
```diff
@@ -92,13 +92,17 @@ struct PackageConfig {
   const PackageType type;
+
+  const HasExports has_exports;
+  const v8::CopyablePersistentTraits<v8::Value>::CopyablePersistent exports;
```
Can you keep using `v8::Global` here, for consistency with the rest of core? There doesn’t seem to be a need for copying semantics here (you do need to use `std::move(exports)` instead of `exports` when initializing the `PackageConfig` below in that case, but that seems to be correct anyway).

jkrems replied:
@guybedford Do you remember why you went with this implementation over `v8::Global`?

guybedford replied:
I think because I didn't know how to use `std::move` so this was my way of working around that initialization failure!?

jkrems replied:
Thanks for the context! I’ll try to make it work with `v8::Global`. 🤞

jkrems replied:
After switching to `v8::Global`, the next issue seems to be that `package_json_cache.emplace` doesn’t like the data structure with a `v8::Global` field. The expansion is pretty long but starts with:

```
/Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:1805:31: error: no
      matching constructor for initialization of 'std::__1::pair<const std::__1::basic_string<char>, const
      node::loader::PackageConfig>'
            ::new((void*)__p) _Up(_VSTD::forward<_Args>(__args)...);
                              ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Not sure yet how that could be resolved.

guybedford replied:
I believe that was exactly the original issue I had.

addaleax replied:
@jkrems Not really sure what to  do with a partial error message here … :smile:

jkrems replied:
@addaleax That’s more than fair. :D I pushed my changes to this PR and also created a gist with the output I get locally: https://gist.github.com/jkrems/7ff6e9b755288f806eefb11010935bc3. I may continue to dig into this later today.

jkrems replied:
Current interpretation:

1. The move constructor provided by `v8::Global<T>` isn’t compatible with the default member initialization generated for structs and/or the struct constructor gets confused by the reference/copy constructor. The end result is that the move constructor of `PackageConfig` gets marked as deleted.
2. I can define a custom move constructor but then I’ll have to also provide a general constructor since it removes the aggregate initialization / brace constructor.

*phew* I’m never gonna underestimate the changes in C++11 again. I’ll go forward with switching this code to an explicit field constructor + move constructor.

jkrems replied:
Okay, turns out that brace-syntax is smart enough to use the constructor with each field spelled out. Neat! I *think* this is all working now. Not 100% sure if implementing them in `env-inl.h` is the appropriate place but that’s where similar code seems to have ended up.

addaleax commented on src/module_wrap.cc:
```diff
@@ -800,6 +804,65 @@ Maybe<URL> PackageMainResolve(Environment* env,
+    Local<String> subpath = String::NewFromUtf8(isolate,
+        pkg_subpath.c_str(), v8::NewStringType::kNormal).ToLocalChecked();
+
+    auto target = exports_obj->Get(context, subpath).ToLocalChecked();
```
Is the `exports` object *guaranteed* to be from JSON at this point (i.e. not a generic user-provided JS object)?

jkrems replied:
I think so, yes. It will always be parsed by the code itself from `package.json`. There shouldn’t be any programmatic way to pass it in.

jkrems commented:
1. Moved to `Global<T>` for `exports`.
2. Fixed type on docs.
3. Added a flag (`--experimental-exports`) that guards this behavior.

The flag is currently in its own commit just in case there's additional discussion over its introduction.

addaleax commented on src/env-inl.h:
```diff
@@ -40,6 +40,31 @@
+      main(pkg.main),
+      type(pkg.type),
+      has_exports(pkg.has_exports),
+      exports((v8::Global<v8::Value>&&) pkg.exports) {
```
style nit: We avoid C-style casts where we can, i.e.

```suggestion
      exports(std::move(pkg.exports)) {
```

addaleax commented on src/env-inl.h:
```diff
@@ -40,6 +40,31 @@
+      main(main),
+      type(type),
+      has_exports(has_exports),
+      exports((v8::Global<v8::Value>&&) exports) {
```
ditto,

```suggestion
      exports(std::move(exports)) {
```

jkrems replied:
I think I had that code initially but ran into a similar "compiler confusion" as with the default implementation, it seems to want to use the deleted reference constructor for some reason:

```
In file included from ../src/base_object-inl.h:28:0,
                 from ../src/async_wrap-inl.h:28,
                 from ../src/api/callback.cc:2:
../src/env-inl.h: In constructor ‘node::loader::PackageConfig::PackageConfig(node::loader::PackageConfig&&)’:
../src/env-inl.h:51:37: error: use of deleted function ‘v8::Global<T>::Global(const v8::Global<T>&) [with T = v8::Value]’
       exports(std::move(pkg.exports)) {
                                     ^
In file included from ../src/node.h:63:0,
                 from ../src/api/callback.cc:1:
../deps/v8/include/v8.h:811:3: note: declared here
   Global(const Global&) = delete;
   ^~~~~~
```

jkrems replied:
Would it help if I made it explicit which cast it is effectively? E.g.:

```cpp
      exports(const_cast<v8::Global<v8::Value>&&>(exports)) {
```

That seems to compile.

addaleax commented on src/env.h:
```diff
@@ -92,13 +92,22 @@ struct PackageConfig {
+  inline PackageConfig(Exists exists, IsValid is_valid, HasMain has_main,
+                       const std::string& main, PackageType type,
+                       HasExports has_exports, v8::Global<v8::Value>&& exports);
+  inline PackageConfig(PackageConfig&& pkg);
```
I’m not sure if you’ve tried it, but if you can get away with 

```suggestion
  PackageConfig(PackageConfig&& pkg) = default;
```

here then I’d do that.

jkrems replied:
I *think* I tried this initially but it still failed but that was a couple of changes ago so I'll definitely double-check. Thanks!

jkrems replied:
Double-checked, running into the deleted constructor with the default implementation (`Global(const Global&) = delete`).

addaleax replied:
@jkrems Yeah, but if you no longer declare `exports` as a const member I think this would go away, plus you can get rid of the invalid `const_cast`s (I didn’t realize before that the casts were necessary because `exports` is defined const, sorry – it’s generally not allowed to cast away constness with the intention of making an object modifiable “by force”).

jkrems replied:
Oh! Let me try what that changes. Maybe that lets us undo a lot of this. 🤞 Thanks for the hint!

jkrems replied:
Ooooh yesssss. Removing `const` from the declaration made all the defaults work which removes the need for a bunch of code to implement custom (and potentially invalid) constructors. Undid the `env-inl.h` changes.

addaleax commented on src/env-inl.h:
```diff
@@ -40,6 +40,31 @@
+inline PackageConfig::PackageConfig(Exists exists, IsValid is_valid,
+                                    HasMain has_main, const std::string& main,
+                                    PackageType type, HasExports has_exports,
+                                    v8::Global<v8::Value>&& exports)
```
```suggestion
                                    v8::Global<v8::Value>&& exports = v8::Global<v8::Value>())
```

I think this should let you get rid of manually using empty values below, if you like.

jkrems replied:
Hm, now I'm tempted to also default `has_exports` to no exports so I can drop both arguments for the common case.

jkrems replied:
Did that which means there's now fewer "random" edits in the other branches.

addaleax commented on src/env.h:
```diff
@@ -92,13 +92,17 @@ struct PackageConfig {
   const std::string main;
   const PackageType type;
+
+  const HasExports has_exports;
```
Ok, one more question and I swear I’m done :smile: Is `has_exports` equivalent to `!exports.IsEmpty()`? In that case this field can be dropped/replaced by a getter, right? :)

jkrems replied:
I think so, yes. My assumption is that @guybedford went with another field either for consistency or because he assumed the check would be more efficient. If there's no efficiency concerns, I think I'd lean towards `IsEmpty()` and drop the second field.

addaleax replied:
@jkrems `IsEmpty()` boils down to testing a single machine word for being zero, so there is most likely no difference in performance and if there is, the benefit is on the side of using `IsEmpty()` because these objects take up less space in total and become more cacheable.

jkrems commented:
(Resolved the threads that still referred to the revision with `const Global<Value> exports`.)

bmeck commented on doc/api/esm.md:
```diff
@@ -216,6 +216,61 @@ a package would be accessible like `require('pkg')` and `import
+`"exports": {}` to indicate the package does not intend for submodules to be
+exposed.
+This is just a convention that works because `false`, just like `{}`, has no
+iterable own properties.
```
can we document what happens if an exports:

* starts with '../'
* starts with '/'
* is an absolute URL
* doesn't lead with './', '../', or '/'
* escapes a package boundary (I actually assume this is allowed looking at the code?)

jkrems replied:
You mean the values/targets of an export mapping? In any case - this is something we want to dig into in a dedicated change for both keys and values. I'd be concerned of missing important cases in the noise if we try to address it in this initial PR. It's listed as one of the blockers to unflagging.

bmeck replied:
seems fine to push off until unflagging.

jkrems commented:
Looks like the Windows test runs are sometimes (?) timing out even though they are still making progress. Retrying the test build directly worked but I'm also trying the top-level fanned build again to see if that's enough. 🤞

jkrems commented:
Landed in 6df7b6a
